clang/lib/CIR/Dialect/Transforms/LoweringPrepare.cpp

```diff
@@ -423,6 +424,137 @@ void LoweringPreparePass::lowerBinOp(BinOp op) {
   op.erase();
 }
 
+static mlir::Value lowerScalarToComplexCast(MLIRContext &ctx, CastOp op) {
+  CIRBaseBuilderTy builder(ctx);
+  builder.setInsertionPoint(op);
+
+  auto src = op.getSrc();
+  auto imag = builder.getNullValue(src.getType(), op.getLoc());
+  return builder.createComplexCreate(op.getLoc(), src, imag);
+}
+
+static mlir::Value lowerComplexToScalarCast(MLIRContext &ctx, CastOp op) {
+  CIRBaseBuilderTy builder(ctx);
+  builder.setInsertionPoint(op);
+
+  auto src = op.getSrc();
+
+  if (!mlir::isa<mlir::cir::BoolType>(op.getType()))
+    return builder.createComplexReal(op.getLoc(), src);
+
+  // Complex cast to bool: (bool)(a+bi) => (bool)a || (bool)b
+  auto srcReal = builder.createComplexReal(op.getLoc(), src);
+  auto srcImag = builder.createComplexImag(op.getLoc(), src);
+
+  mlir::cir::CastKind elemToBoolKind;
+  if (op.getKind() == mlir::cir::CastKind::float_complex_to_bool)
+    elemToBoolKind = mlir::cir::CastKind::float_to_bool;
+  else if (op.getKind() == mlir::cir::CastKind::int_complex_to_bool)
+    elemToBoolKind = mlir::cir::CastKind::int_to_bool;
+  else
+    llvm_unreachable("invalid complex to bool cast kind");
+
+  auto boolTy = builder.getBoolTy();
+  auto srcRealToBool =
+      builder.createCast(op.getLoc(), elemToBoolKind, srcReal, boolTy);
+  auto srcImagToBool =
+      builder.createCast(op.getLoc(), elemToBoolKind, srcImag, boolTy);
+
+  // srcRealToBool || srcImagToBool
+  return builder
+      .create<mlir::cir::TernaryOp>(
+          op.getLoc(), srcRealToBool,
+          [&](mlir::OpBuilder &, mlir::Location) {
+            builder.createYield(op.getLoc(),
+                                builder.getTrue(op.getLoc()).getResult());
+          },
+          [&](mlir::OpBuilder &, mlir::Location) {
+            auto inner =
+                builder
+                    .create<mlir::cir::TernaryOp>(
+                        op.getLoc(), srcImagToBool,
+                        [&](mlir::OpBuilder &, mlir::Location) {
+                          builder.createYield(
+                              op.getLoc(),
+                              builder.getTrue(op.getLoc()).getResult());
+                        },
+                        [&](mlir::OpBuilder &, mlir::Location) {
+                          builder.createYield(
+                              op.getLoc(),
+                              builder.getFalse(op.getLoc()).getResult());
+                        })
+                    .getResult();
+            builder.createYield(op.getLoc(), inner);
+          })
+      .getResult();
```

Comment:
This is not the first time I write code like these to implement a simple boolean binary op like `a && b` or `a || b`. I think a dedicated operation for binary logical operations are really necessary and they allow generation of much better LLVM IR, although these operations won't be generated from CIRGen directly.

Reply:
I have gone through the same feeling, we need a ternary that's basically a select operation. If you wanna add that it would be a great addition. `cir.ternary_select` or something, then a fold operation can even transform a `cir.ternary` into `cir.ternary_select` to handle all really simple cases like this one (some CIRGen will use generic code and might emit this pattern regardless of cases we actually know we want a select-like op.

Reply:
> I have gone through the same feeling, we need a ternary that's basically a select operation.

Do we additionally need a boolean binary op, something like `cir.binop logical_and` ?

Reply:
I'm not sure how usable this would be, given that usually those have short-circuiting semantics. We have folders and other transformations passes that can help out with this, have you explored any of those solutions already?

Reply:
Let's track this in #785 .